src/index.ts

```diff
@@ -0,0 +1,522 @@
+      const error: Error = Error(errorEvent["message"]);
+      error["stack"] =
+        "Error: " +
+        errorEvent["message"] +
+        "\n    at " +
+        errorEvent["filename"] +
+        ":" +
+        errorEvent["lineno"] +
+        ":" +
+        errorEvent["colno"];
+      resolveModule({ id, state: ModuleState.ERROR, error });
+    };
+    doc.head.appendChild(el);
+    return {
+      id,
+      requestId,
+      state: ModuleState.NETWORK_LOADING,
+      callbacks: [callback]
+    };
+  }
+
+  function throwError(error: Error) {
+    throw error;
+  }
+
+  function defer<T extends any[]>(fn: (...args: T) => void, args: T) {
+    setTimeout(function() {
+      fn.apply(undefined, args);
+    });
+  }
+
+  function doDefine(
+    id: ModuleId,
+    dependencies: string[],
+    factory: IamdeeDefineFactoryFunction
+  ) {
+    const existingModule = moduleMap.get(id);
+    const definedModule: DefinedModule = {
+      state: ModuleState.DEFINED,
+      id,
+      callbacks: [],
+      load(requestId) {
+        const waitingModule: WaitingForDependenciesModule = {
+          id,
+          requestId,
+          state: ModuleState.WAITING_FOR_DEPENDENCIES,
+          callbacks: definedModule.callbacks,
+          exports: {}
+        };
+        moduleMap.set(id, waitingModule);
+        const localRequire = createRequire(id, requestId);
+        localRequire(
+          dependencies,
+          function() {
+            const result = factory.apply(undefined, arguments);
+            const exports =
+              result === undefined ? waitingModule.exports : result;
+            resolveModule({ id, state: ModuleState.INITIALIZED, exports });
+          },
+          function(error) {
+            resolveModule({ id, state: ModuleState.ERROR, error });
+          }
+        );
+      }
+    };
+    moduleMap.set(id, definedModule);
+    if (existingModule) {
+      if (existingModule.state != ModuleState.NETWORK_LOADING) {
+        return panic("Trying to define a module that is in a wrong state");
+      }
+      definedModule.callbacks = existingModule.callbacks;
+      definedModule.load(existingModule.requestId);
+    }
+  }
+
+  // This is a live node list, so we do not need to re-query
+  const scripts = doc.getElementsByTagName("script");
+
+  function getCurrentScript() {
+    let script = doc["currentScript"];
+    if (script) {
+      return script;
+    }
+    // Support for IE 9-11
+    try {
+      throw Error();
+    } catch (err) {
+      // Internet Explorer 11 does not support currentScript,
+      // so we are inspecting error stack for the right script url
+      const lastLine = err.stack ? err.stack.split("\n").pop() : "";
+      for (let i = scripts.length - 1; i >= 0; --i) {
+        script = scripts[i];
+        if (script.src) {
+          if (lastLine.indexOf(script.src) >= 0) {
+            return script;
+          } else if ((script as any).readyState == "interactive") {
```

Comment:
Should check for readyState support in script. don't think its supported.

Reply:
Checking for support is just `"readyState" in script`. Considering that I'm just reading the property, why would that be safer than current comparison? Am I missing something?